File: api/Services/SonosCommandService.cs
using System.Text.Json;
using SonosSoundHub.Models;

namespace SonosSoundHub.Services;

/// <summary>
/// Service to execute commands via the soco-cli HTTP API.
/// Uses a semaphore to serialize requests - soco-cli cannot handle concurrent requests properly.
/// </summary>
public class SonosCommandService
{
    private readonly HttpClient _httpClient;
    private readonly SocoCliService _socoCliService;
    private readonly ILogger<SonosCommandService> _logger;
    
    /// <summary>
    /// Semaphore to ensure only one request to soco-cli at a time.
    /// soco-cli's HTTP API can mix up responses when handling concurrent requests.
    /// </summary>
    private static readonly SemaphoreSlim _requestLock = new(1, 1);

    private static async Task<string> ReadBodySafeAsync(HttpResponseMessage response)
    {
        try
        {
            return await response.Content.ReadAsStringAsync();
        }
        catch
        {
            return "<unable to read response body>";
        }
    }

    public SonosCommandService(
        HttpClient httpClient,
        SocoCliService socoCliService,
        ILogger<SonosCommandService> logger)
    {
        _httpClient = httpClient;
        _socoCliService = socoCliService;
        _logger = logger;
    }

    /// <summary>
    /// Gets the list of speakers
    /// </summary>
    public async Task<List<string>> GetSpeakersAsync()
    {
        await _socoCliService.EnsureServerRunningAsync();
        await _requestLock.WaitAsync();

        try
        {
            var url = $"{_socoCliService.ServerUrl}/speakers";
            var response = await _httpClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
            {
                var body = await ReadBodySafeAsync(response);
                _logger.LogError(
                    "soco-cli request failed: GET {Url} => {StatusCode} {ReasonPhrase}. Body: {Body}",
                    url,
                    (int)response.StatusCode,
                    response.ReasonPhrase,
                    body);
                return new List<string>();
            }

            var content = await response.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<JsonElement>(content);
            
            if (result.TryGetProperty("speakers", out var speakers))
            {
                return speakers.EnumerateArray()
                    .Select(s => s.GetString() ?? string.Empty)
                    .Where(s => !string.IsNullOrEmpty(s))
                    .ToList();
            }

            return new List<string>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get speakers");
            return new List<string>();
        }
        finally
        {
            _requestLock.Release();
        }
    }

    /// <summary>
    /// Triggers speaker rediscovery
    /// </summary>
    public async Task<List<string>> RediscoverSpeakersAsync()
    {
        await _socoCliService.EnsureServerRunningAsync();
        await _requestLock.WaitAsync();

        try
        {
            var url = $"{_socoCliService.ServerUrl}/rediscover";
            var response = await _httpClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
            {
                var body = await ReadBodySafeAsync(response);
                _logger.LogError(
                    "soco-cli request failed: GET {Url} => {StatusCode} {ReasonPhrase}. Body: {Body}",
                    url,
                    (int)response.StatusCode,
                    response.ReasonPhrase,
                    body);
                return new List<string>();
            }

            var content = await response.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<JsonElement>(content);
            
            if (result.TryGetProperty("speakers_discovered", out var speakers))
            {
                return speakers.EnumerateArray()
                    .Select(s => s.GetString() ?? string.Empty)
                    .Where(s => !string.IsNullOrEmpty(s))
                    .ToList();
            }

            return new List<string>();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to rediscover speakers");
            return new List<string>();
        }
        finally
        {
            _requestLock.Release();
        }
    }

    /// <summary>
    /// Executes a command on a speaker
    /// </summary>
    public async Task<SocoCliResponse> ExecuteCommandAsync(string speaker, string action, params string[] args)
    {
        await _socoCliService.EnsureServerRunningAsync();
        await _requestLock.WaitAsync();

        try
        {
            var url = $"{_socoCliService.ServerUrl}/{Uri.EscapeDataString(speaker)}/{Uri.EscapeDataString(action)}";
            
            if (args.Length > 0)
            {
                var encodedArgs = args.Select(Uri.EscapeDataString);
                url += "/" + string.Join("/", encodedArgs);
            }

            _logger.LogDebug("Executing command: {Url}", url);

            var response = await _httpClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
            {
                var body = await ReadBodySafeAsync(response);
                _logger.LogError(
                    "soco-cli request failed: GET {Url} => {StatusCode} {ReasonPhrase}. Body: {Body}",
                    url,
                    (int)response.StatusCode,
                    response.ReasonPhrase,
                    body);
                return new SocoCliResponse
                {
                    Speaker = speaker,
                    Action = action,
                    Args = args,
                    ExitCode = (int)response.StatusCode,
                    ErrorMsg = $"HTTP {(int)response.StatusCode} {response.ReasonPhrase}"
                };
            }

            var content = await response.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<SocoCliResponse>(content, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            });

            return result ?? new SocoCliResponse { ErrorMsg = "Failed to parse response" };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to execute command: {Speaker} {Action}", speaker, action);
            return new SocoCliResponse
            {
                Speaker = speaker,
                Action = action,
                Args = args,
                ExitCode = -1,
                ErrorMsg = ex.Message
            };
        }
        finally
        {
            _requestLock.Release();
        }
    }

    /// <summary>
    /// Checks if an error message indicates the speaker is offline/unreachable
    /// </summary>
    private static bool IsTimeoutOrConnectionError(string? errorMsg)
    {
        if (string.IsNullOrEmpty(errorMsg))
            return false;

        var lowerError = errorMsg.ToLowerInvariant();
        return lowerError.Contains("timed out") ||
               lowerError.Contains("timeout") ||
               lowerError.Contains("connection refused") ||
               lowerError.Contains("unreachable") ||
               lowerError.Contains("no route to host") ||
               lowerError.Contains("network is unreachable") ||
               lowerError.Contains("connecttimeouterror") ||
               lowerError.Contains("max retries exceeded");
    }

    /// <summary>
    /// Gets detailed information about a speaker
    /// </summary>
    public async Task<Speaker> GetSpeakerInfoAsync(string speakerName)
    {
        var speaker = new Speaker { Name = speakerName };

        try
        {
            // Get volume first as a connectivity check
            var volumeResponse = await ExecuteCommandAsync(speakerName, "volume");
            
            // Check if this is a timeout/connection error indicating offline speaker
            if (volumeResponse.ExitCode != 0 && IsTimeoutOrConnectionError(volumeResponse.ErrorMsg))
            {
                speaker.IsOffline = true;
                speaker.ErrorMessage = "Speaker is offline or unreachable";
                _logger.LogWarning("Speaker {Speaker} appears to be offline: {Error}", speakerName, volumeResponse.ErrorMsg);
                return speaker;
            }
            
            if (volumeResponse.ExitCode == 0 && int.TryParse(volumeResponse.Result, out var volume))
            {
                speaker.Volume = volume;
            }

            // Get mute status
            var muteResponse = await ExecuteCommandAsync(speakerName, "mute");
            if (muteResponse.ExitCode != 0 && IsTimeoutOrConnectionError(muteResponse.ErrorMsg))
            {
                speaker.IsOffline = true;
                speaker.ErrorMessage = "Speaker is offline or unreachable";
                return speaker;
            }
            if (muteResponse.ExitCode == 0)
            {
                speaker.IsMuted = muteResponse.Result.ToLower() == "on";
            }

            // Get playback state
            var stateResponse = await ExecuteCommandAsync(speakerName, "playback");
            if (stateResponse.ExitCode != 0 && IsTimeoutOrConnectionError(stateResponse.ErrorMsg))
            {
                speaker.IsOffline = true;
                speaker.ErrorMessage = "Speaker is offline or unreachable";
                return speaker;
            }
            if (stateResponse.ExitCode == 0)
            {
                speaker.PlaybackState = stateResponse.Result;
            }

            // Get current track
            var trackResponse = await ExecuteCommandAsync(speakerName, "track");
            if (trackResponse.ExitCode != 0 && IsTimeoutOrConnectionError(trackResponse.ErrorMsg))
            {
                speaker.IsOffline = true;
                speaker.ErrorMessage = "Speaker is offline or unreachable";
                return speaker;
            }
            if (trackResponse.ExitCode == 0)
            {
                speaker.CurrentTrack = trackResponse.Result;
            }

            // Get battery level (for portable speakers like Roam/Move)
            var batteryResponse = await ExecuteCommandAsync(speakerName, "battery");
            if (batteryResponse.ExitCode == 0 && !string.IsNullOrEmpty(batteryResponse.Result))
            {
                // Battery response format is typically just a number like "85"
                if (int.TryParse(batteryResponse.Result.Trim().Replace("%", ""), out var batteryLevel))
                {
                    speaker.BatteryLevel = batteryLevel;
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get speaker info for {Speaker}", speakerName);
            
            // Check if the exception indicates a timeout/connection issue
            if (IsTimeoutOrConnectionError(ex.Message))
            {
                speaker.IsOffline = true;
                speaker.ErrorMessage = "Speaker is offline or unreachable";
            }
        }

        return speaker;
    }
}

File: api-python/src/sonos_hub/services/sonos_command_service.py
"""Service to execute commands via the soco-cli HTTP API."""

import asyncio
import logging
from typing import Any
from urllib.parse import quote

import httpx

from ..config import Settings
from ..models import Speaker, SocoCliResponse
from .soco_cli_service import SocoCliService

logger = logging.getLogger(__name__)


class SonosCommandService:
    """Service to execute commands via the soco-cli HTTP API.
    
    Uses a semaphore to serialize requests - soco-cli cannot handle concurrent requests properly.
    """
    
    def __init__(self, settings: Settings, soco_cli_service: SocoCliService):
        """Initialize the service.
        
        Args:
            settings: Application settings.
            soco_cli_service: The soco-cli service.
        """
        self._settings = settings
        self._soco_cli_service = soco_cli_service
        self._request_lock = asyncio.Lock()
        self._client: httpx.AsyncClient | None = None
    
    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the HTTP client."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(timeout=30.0)
        return self._client
    
    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client:
            await self._client.aclose()
            self._client = None
    
    async def get_speakers(self) -> list[str]:
        """Get the list of speakers.
        
        Returns:
            List of speaker names.
        """
        await self._soco_cli_service.ensure_server_running()
        async with self._request_lock:
            try:
                client = await self._get_client()
                url = f"{self._soco_cli_service.server_url}/speakers"
                response = await client.get(url)
                
                if response.status_code != 200:
                    logger.error(
                        "soco-cli request failed: GET %s => %d. Body: %s",
                        url, response.status_code, response.text
                    )
                    return []
                
                result = response.json()
                if "speakers" in result:
                    return [s for s in result["speakers"] if s]
                return []
                
            except Exception as e:
                logger.error("Failed to get speakers: %s", e)
                return []
    
    async def rediscover_speakers(self) -> list[str]:
        """Trigger speaker rediscovery.
        
        Returns:
            List of discovered speaker names.
        """
        await self._soco_cli_service.ensure_server_running()
        async with self._request_lock:
            try:
                client = await self._get_client()
                url = f"{self._soco_cli_service.server_url}/rediscover"
                response = await client.get(url)
                
                if response.status_code != 200:
                    logger.error(
                        "soco-cli request failed: GET %s => %d. Body: %s",
                        url, response.status_code, response.text
                    )
                    return []
                
                result = response.json()
                if "speakers_discovered" in result:
                    return [s for s in result["speakers_discovered"] if s]
                return []
                
            except Exception as e:
                logger.error("Failed to rediscover speakers: %s", e)
                return []
    
    async def execute_command(
        self, speaker: str, action: str, *args: str
    ) -> SocoCliResponse:
        """Execute a command on a speaker.
        
        Args:
            speaker: Speaker name.
            action: Command action.
            args: Additional arguments.
            
        Returns:
            Response from soco-cli.
        """
        await self._soco_cli_service.ensure_server_running()
        async with self._request_lock:
            try:
                client = await self._get_client()
                url = f"{self._soco_cli_service.server_url}/{quote(speaker)}/{quote(action)}"
                
                if args:
                    encoded_args = "/".join(quote(arg) for arg in args)
                    url += f"/{encoded_args}"
                
                logger.debug("Executing command: %s", url)
                
                response = await client.get(url)
                
                if response.status_code != 200:
                    logger.error(
                        "soco-cli request failed: GET %s => %d. Body: %s",
                        url, response.status_code, response.text
                    )
                    return SocoCliResponse(
                        speaker=speaker,
                        action=action,
                        args=list(args),
                        exit_code=response.status_code,
                        error_msg=f"HTTP {response.status_code}",
                    )
                
                result = response.json()
                return SocoCliResponse(
                    speaker=result.get("speaker", speaker),
                    action=result.get("action", action),
                    args=result.get("args", list(args)),
                    exit_code=result.get("exit_code", 0),
                    result=result.get("result", ""),
                    error_msg=result.get("error_msg", ""),
                )
                
            except Exception as e:
                logger.error("Failed to execute command %s %s: %s", speaker, action, e)
                return SocoCliResponse(
                    speaker=speaker,
                    action=action,
                    args=list(args),
                    exit_code=-1,
                    error_msg=str(e),
                )
    
    @staticmethod
    def _is_timeout_or_connection_error(error_msg: str | None) -> bool:
        """Check if an error message indicates the speaker is offline/unreachable."""
        if not error_msg:
            return False
        
        lower_error = error_msg.lower()
        return any(s in lower_error for s in [
            "timed out", "timeout", "connection refused", "unreachable",
            "no route to host", "network is unreachable", "connecttimeouterror",
            "max retries exceeded",
        ])
    
    async def get_speaker_info(self, speaker_name: str) -> Speaker:
        """Get detailed information about a speaker.
        
        Args:
            speaker_name: Name of the speaker.
            
        Returns:
            Speaker information.
        """
        speaker = Speaker(name=speaker_name)
        
        try:
            # Get volume first as a connectivity check
            volume_response = await self.execute_command(speaker_name, "volume")
            
            if volume_response.exit_code != 0 and self._is_timeout_or_connection_error(volume_response.error_msg):
                speaker.is_offline = True
                speaker.error_message = "Speaker is offline or unreachable"
                logger.warning(
                    "Speaker %s appears to be offline: %s",
                    speaker_name, volume_response.error_msg
                )
                return speaker
            
            if volume_response.exit_code == 0:
                try:
                    speaker.volume = int(volume_response.result)
                except ValueError:
                    pass
            
            # Get mute status
            mute_response = await self.execute_command(speaker_name, "mute")
            if mute_response.exit_code != 0 and self._is_timeout_or_connection_error(mute_response.error_msg):
                speaker.is_offline = True
                speaker.error_message = "Speaker is offline or unreachable"
                return speaker
            if mute_response.exit_code == 0:
                speaker.is_muted = mute_response.result.lower() == "on"
            
            # Get playback state
            state_response = await self.execute_command(speaker_name, "playback")
            if state_response.exit_code != 0 and self._is_timeout_or_connection_error(state_response.error_msg):
                speaker.is_offline = True
                speaker.error_message = "Speaker is offline or unreachable"
                return speaker
            if state_response.exit_code == 0:
                speaker.playback_state = state_response.result
            
            # Get current track
            track_response = await self.execute_command(speaker_name, "track")
            if track_response.exit_code != 0 and self._is_timeout_or_connection_error(track_response.error_msg):
                speaker.is_offline = True
                speaker.error_message = "Speaker is offline or unreachable"
                return speaker
            if track_response.exit_code == 0:
                speaker.current_track = track_response.result
            
            # Get battery level (for portable speakers like Roam/Move)
            battery_response = await self.execute_command(speaker_name, "battery")
            if battery_response.exit_code == 0 and battery_response.result:
                try:
                    battery_str = battery_response.result.strip().replace("%", "")
                    speaker.battery_level = int(battery_str)
                except ValueError:
                    pass
                    
        except Exception as e:
            logger.error("Failed to get speaker info for %s: %s", speaker_name, e)
            speaker.error_message = str(e)
        
        return speaker
